e2e/testcases/basic_test.go

```diff
@@ -0,0 +1,324 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	v1 "kpt.dev/configsync/pkg/api/configmanagement/v1"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/testing/fake"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	yamlDir = "../testdata"
+)
+
+func TestNoDefaultFieldsInNamespaceConfig(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.Reconciliation2, ntopts.SkipMultiRepo)
+	expectedName := "hello-world"
+	expectedTime := metav1.Time{}
+
+	nt.T.Log("Add a deployment")
+	nt.RootRepos[configsync.RootSyncName].Copy(fmt.Sprintf("%s/dir-namespace.yaml", yamlDir), "acme/namespaces/dir/namespace.yaml")
+	nt.RootRepos[configsync.RootSyncName].Copy(fmt.Sprintf("%s/deployment-helloworld.yaml", yamlDir), "acme/namespaces/dir/deployment.yaml")
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Adding deployment")
+	nt.WaitForRepoSyncs()
+
+	nt.T.Log("Check that the deployment was created")
+	if err := nt.Validate("hello-world", "dir", &appsv1.Deployment{}); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.T.Log("Check that specified field name is present and unspecified field creationTimestamp is not present")
+	if err := nt.Validate("dir", "", &v1.NamespaceConfig{},
+		func(o client.Object) error {
+			raw := o.(*v1.NamespaceConfig).Spec.Resources[0].Versions[0].Objects[0].Raw
+			dep := &appsv1.Deployment{}
+			if err := json.Unmarshal(raw, &dep); err != nil {
+				return errors.Errorf("Failed to get raw JSON, %s", err)
+			}
+			actualName := dep.GetName()
+			if actualName != expectedName {
+				return errors.Errorf("Expected name: %s, got: %s", expectedName, actualName)
+			}
+			actualTime := dep.GetCreationTimestamp()
+			if actualTime != expectedTime {
```

Comment:
Just noticed that you're doing an equality check between the actual creation time and the expected time (the current time), which will make the test extremely flaky and may even fail on autopilot clusters. The creationTime is controller by the applier.

Instead of an equality check, you can check if the actualTime is no earlier than the expected time.

Reply:
From the original bats test, creationTimestamp is not supposed to exist in the metadata, therefore when the test do `kubectl get namespaceconfig dir -ojson | jq -c .spec.resources[0].versions[0].objects[0].metadata | grep creationTimestamp`, it will simply returns empty. I represent this in Golang by creating an empty time object in `expectedTime` and getting the creationTimestamp from `GetCreationTimestamp()` which should also return an empty time object. Do you think i can make an improvement to the comparison by checking the actualTime is no earlier than the expected time?

Reply:
ahh, I see. The expected time is not set to current, but empty. I think the current approach looks good.